packages/dds/tree/src/shared-tree/defaultCommitEnricher.ts

```diff
@@ -0,0 +1,176 @@
+	}
+
+	public enrichCommit(commit: GraphCommit<TChange>): GraphCommit<TChange> {
+		const toResubmit = this.resubmitQueue.shift();
+		if (toResubmit !== undefined) {
+			this.inFlightQueue.push(toResubmit);
+			return toResubmit;
+		} else {
+			const updatedChange = this.tip.updateChangeEnrichments(commit.change, commit.revision);
+			const updatedCommit = { ...commit, change: updatedChange };
+			this.inFlightQueue.push(updatedCommit);
+			this.tip[disposeSymbol]();
+			this.tip = this.checkoutFactory();
+			return updatedCommit;
+		}
+	}
+
+	public prepareForResubmit(toResubmit: readonly GraphCommit<TChange>[]): void {
+		assert(
+			!this.isInResubmitPhase,
+			"Invalid resubmit phase start during incomplete resubmit phase",
+		);
+		assert(
+			toResubmit.length === this.inFlightQueue.length,
+			"Unexpected resubmit of more or fewer commits than are in flight",
+		);
+		if (this.latestInFlightCommitWithStaleEnrichments === -1) {
+			// No in-flight commits have stale enrichments, so we can resubmit them as is
+			this.resubmitQueue = this.inFlightQueue;
+			this.inFlightQueue = [];
+		} else {
+			const checkout = this.checkoutFactory();
+			// Roll back the checkout to the state before the oldest commit
+			for (let iCommit = toResubmit.length - 1; iCommit >= 0; iCommit -= 1) {
+				const commit = toResubmit[iCommit];
+				const rollback = this.inverter(commit, true);
+				// WARNING: it's not currently possible to roll back past a schema change (see AB#7265).
+				// Either we have to make it possible to do so, or this logic will have to change to work
+				// forwards from an earlier fork instead of backwards.
+				checkout.applyTipChange(rollback);
+			}
+			// Update the enrichments of the stale commits
+			for (
+				let iCommit = 0;
+				iCommit <= this.latestInFlightCommitWithStaleEnrichments;
+				iCommit += 1
+			) {
+				const commit = toResubmit[iCommit];
+				const enrichedChange = checkout.updateChangeEnrichments(
+					commit.change,
+					commit.revision,
+				);
+				const enrichedCommit = { ...commit, change: enrichedChange };
+				this.resubmitQueue.push(enrichedCommit);
+				if (iCommit < this.latestInFlightCommitWithStaleEnrichments) {
+					checkout.applyTipChange(enrichedChange, commit.revision);
+				}
+				this.inFlightQueue.shift();
+			}
+			// Whatever commits are left do not have stale enrichments
+			for (const commit of this.inFlightQueue) {
+				this.resubmitQueue.push(commit);
+			}
+			this.inFlightQueue.length = 0;
+		}
+		this.latestInFlightCommitWithStaleEnrichments = -1;
+	}
+
+	public get isInResubmitPhase(): boolean {
+		return this.resubmitQueue.length !== 0;
+	}
+
+	public onSequencedCommitApplied(isLocal: boolean): void {
+		if (isLocal) {
+			// The oldest in-flight commit has been sequenced
+			assert(this.inFlightQueue.length > 0, "Sequencing of unknown local commit");
+			this.inFlightQueue.shift();
+			if (this.latestInFlightCommitWithStaleEnrichments >= 0) {
```

Comment:
nit: You can just remove this check altogether, because if this.latest negative then it is already exactly `-1` and setting it again will do no harm.

Reply:
But we don't set it to `-1` here, we only decrement it. So if it were already `-1` it would become `-2`.

Reply:
Oops! I really misread that. 

= -1 != -= 1

as they say